ping.go

```diff
@@ -407,51 +408,62 @@ func (p *Pinger) Run() error {
 		handler()
 	}
 
+	timeout := time.NewTicker(p.Timeout)
+	interval := time.NewTicker(p.Interval)
+	defer func() {
+		p.Stop()
+		interval.Stop()
+		timeout.Stop()
+		wg.Wait()
+	}()
+
 	err = p.sendICMP(conn)
 	if err != nil {
 		return err
 	}
 
-	timeout := time.NewTicker(p.Timeout)
-	defer timeout.Stop()
-	interval := time.NewTicker(p.Interval)
-	defer interval.Stop()
-
 	for {
 		select {
 		case <-p.done:
-			wg.Wait()
 			return nil
 		case <-timeout.C:
-			close(p.done)
-			wg.Wait()
 			return nil
+		case r := <-recv:
+			err := p.processPacket(r)
+			if err != nil {
+				// FIXME: this logs as FATAL but continues
+				fmt.Println("FATAL:", err)
+			}
 		case <-interval.C:
 			if p.Count > 0 && p.PacketsSent >= p.Count {
+				interval.Stop()
 				continue
 			}
 			err = p.sendICMP(conn)
 			if err != nil {
 				// FIXME: this logs as FATAL but continues
-				fmt.Println("FATAL: ", err.Error())
-			}
-		case r := <-recv:
-			err := p.processPacket(r)
-			if err != nil {
-				// FIXME: this logs as FATAL but continues
-				fmt.Println("FATAL: ", err.Error())
+				fmt.Println("FATAL:", err)
 			}
 		}
 		if p.Count > 0 && p.PacketsRecv >= p.Count {
-			close(p.done)
-			wg.Wait()
 			return nil
 		}
 	}
 }
 
 func (p *Pinger) Stop() {
-	close(p.done)
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	open := true
+	select {
+	case _, open = <-p.done:
```

Comment:
the p.done channel has been changed to `chan interface{}` right? I actually didn't think that this would compile and Im not entirely sure I understand the intended behavior. You are setting a boolean variable to `interface{}`? Does that make it `false`?

Reply:
nevermind, `_` is the interface{}, i see now